### ISSUE TITLE
SCRAM-SHA-256(-PLUS) supports

### ISSUE THREAD

**Issue by Neustradamus** [DbUp/DbUp#599]
Dear @DbUp team,

In first, I wish you a Happy New Year!

Can you add supports of:
- SCRAM-SHA-256
- SCRAM-SHA-256-PLUS

Thanks in advance.

All informations about SCRAM:

"When using the SASL SCRAM mechanism, the SCRAM-SHA-256-PLUS variant SHOULD be preferred over the SCRAM-SHA-256 variant, and SHA-256 variants [RFC7677] SHOULD be preferred over SHA-1 variants [RFC5802]".

- SCRAM-SHA-1(-PLUS):
-- https://tools.ietf.org/html/rfc5802
-- https://tools.ietf.org/html/rfc6120

- SCRAM-SHA-256(-PLUS):
-- https://tools.ietf.org/html/rfc7677 since 2015-11-02
-- https://tools.ietf.org/html/rfc8600 since 2019-06-21: https://mailarchive.ietf.org/arch/msg/ietf-announce/suJMmeMhuAOmGn_PJYgX5Vm8lNA

- SCRAM-SHA-512(-PLUS):
-- https://tools.ietf.org/html/draft-melnikov-scram-sha-512

- SCRAM-SHA3-512(-PLUS):
-- https://tools.ietf.org/html/draft-melnikov-scram-sha3-512

https://xmpp.org/extensions/inbox/hash-recommendations.html

-PLUS variants:
- RFC5056: On the Use of Channel Bindings to Secure Channels: https://tools.ietf.org/html/rfc5056
- RFC5929: Channel Bindings for TLS: https://tools.ietf.org/html/rfc5929
- Channel-Binding Types: https://www.iana.org/assignments/channel-binding-types/channel-binding-types.xhtml
- RFC 9266: Channel Bindings for TLS 1.3: https://tools.ietf.org/html/rfc9266

IMAP:
- RFC9051: Internet Message Access Protocol (IMAP) - Version 4rev2: https://tools.ietf.org/html/rfc9051

LDAP:
- RFC5803: Lightweight Directory Access Protocol (LDAP) Schema for Storing Salted: Challenge Response Authentication Mechanism (SCRAM) Secrets: https://tools.ietf.org/html/rfc5803

HTTP:
- RFC7804: Salted Challenge Response HTTP Authentication Mechanism: https://tools.ietf.org/html/rfc7804

2FA:
- Extensions to Salted Challenge Response (SCRAM) for 2 factor authentication: https://tools.ietf.org/html/draft-melnikov-scram-2fa

IANA:
- Simple Authentication and Security Layer (SASL) Mechanisms: https://www.iana.org/assignments/sasl-mechanisms/sasl-mechanisms.xhtml

Linked to:
- https://github.com/scram-xmpp/info/issues/1

**Comment by AdrianJSClark** [comment]
DbUp doesn't implement any authentication or security protocols, @Neustradamus. DbUp will use the protocols which are implemented by the libraries that it uses for connectivity to the various database servers.

**Comment by huesie** [comment]
@Neustradamus thanks for nice detailed info.  DbUp/dbup-postgresql includes Npgsql v3.2.**6** and SCRAM authentication is supported in Npgsql v3.2.**7**+.  Specifically SCRAM-SHA-256, as detailed in error messages or alternatively "Authentication method not supported (Received: 10)" is fixed by using Npgsql v3.2.7 or later.

I'll raise a PR if someone is kind enough to approve this - there are lots of later versions of Npgsql to choose from v4.x; v5.x; and v6.x but I've chosen to upgrade from v3.2.6→v3.2.7 as I've verified that this works - just in case someone still requires .NET Standard v1.3.

**Comment by Neustradamus** [comment]
It is official for TLS 1.3 Binding!
- RFC 9266: Channel Bindings for TLS 1.3: https://tools.ietf.org/html/rfc9266

Details:
- tls-unique for TLS =< 1.2
- tls-exporter for TLS = 1.3

**Comment by mjauernig** [comment]
Seems to be done already. If still needed, please reopen.

**Comment by Neustradamus** [comment]
@mjauernig: Can you reopen?
- https://github.com/search?q=repo%3ADbUp%2FDbUp+tls-exporter&type=code
- https://github.com/search?q=repo%3ADbUp%2FDbUp+tls-unique&type=code

Nothing.

**Comment by mjauernig** [comment]
@Neustradamus 
Yes, but can you please give a deeper hint of what is missing. What do we need to implement?

@huesie wrote 
> @Neustradamus thanks for nice detailed info. DbUp/dbup-postgresql includes Npgsql v3.2.6 and SCRAM authentication is supported in Npgsql v3.2.7+. Specifically SCRAM-SHA-256, as detailed in error messages or alternatively "Authentication method not supported (Received: 10)" is fixed by using Npgsql v3.2.7 or later.

The dependeny is on 3.2.7
>  <PackageReference Include="Npgsql" Version="3.2.7" />

So it looks resolved on a simple first look.
